### ISSUE TITLE
High level expression optimization

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#4038]
### Summary

We may be able to provide high-level expression optimization within Dask by extending the current ShareDict system with a family of custom mutable mappings.  We're experimenting with this in dask array and it is showing promise.  This might motivate a larger change to the Dask collections.

### Problem

Currently dask collections provide no mechanism to do high level query optimizations.  

Some examples of high level query optimizations that are commonly requested:

-  Fuse elementwise operations

    ```python
    Before: (x + 1) + 2)
    After: x + (1 + 2)
    ```
-   Move slicing operations before elementwise operations

    ```python
    Before:(x + 1)[:5]
    After: x[:5] + 1
    ```

-   Or only read a few columns out of a parquet dataset based on use

    ```python
    Before: df = dd.read_parquet('my-dataset.parquet')
    Before: df.x.sum().compute()

    After: df = dd.read_parquet('my-dataset.parquet', columns=['x'])
    After: df.x.sum().compute()
    ```
-  Or in a more grand scale, optimize large complex dataframe comptuations with joins, filters, projections, etc.

This is hard today because dask collections generate low-level task graphs immediately.  This approach is good for flexibility (we are not restricted to making special classes of operations), but makes it very difficult to reason about our computation at a high level.  There is some tension between supporting any arbitrary dynamic task graph, and supporting a high-level alegbra of operations that we know how to rearrange.

### But in practice there is high level structure to our graphs

For the high-level collections (dask array/bag/dataframe) the task graph is well organized into layers of tasks that tend to only depend on a small number of other layers.  If we wanted to we could probably back out a high level graph.  

These layers are explicit in the use of `ShareDict` in dask.array which, in an original effort to minimize duplication of large graphs, stores each layer separately in a dict-of-dicts.  The layers in a ShareDict are precisely the higher level structure, except that we lack two things:

1.  Annotation about where they came from.  Are they an elementwise computation?  A selection?  A join?
2.  Dependencies between the layers

When trying to implement a bit of high level optimization for dask array atop operations in #3998 we ended up implementing the above for dask array for one broad class of operations.  So far this hasn't been *too* painful.  

### Proposal

So, we might consider the following:

1.  Use ShareDict more liberally, specifically in dask.dataframe and more in base.py
2.  Replace *some* of the dictionaries we insert in dask computations with lazily evaluated Mappings that include symbolic information
3.  Build optimations around Sharedicts to manipulate them a bit before lowering the custom Mappings to a normal dict and then doing task-based optimizations

**Comment by mrocklin** [comment]
I had a conversation with @jcirst on this topic.  Here are a few points:

1.  It may eventually be useful to have a separate optimization phase that includes information about the collections.  Today this would happen as an extra call in `dask/base.py::collections_to_dsk` to some new protocol function.
2.  This would allow us to avoid optimization at boundaries between collections like `DataFrame.to_dask_array`, and instead optimize the entire graph based on regions of different collections.
3.  We can probably get atop fusion in before doing this
4.  We can probabl also incrementally get things like Bag and DataFrame over to use ShareDict (need a better name?) with dependencies
5.  We might also look at visualization of the higher level graph.
6.  Optimizations are likely to be strictly separated between high level optimizations (such as those decribed here) followed by a strict conversion to Python's builtin `dict`, followed by low-level optimizations (`cull`, `fuse`, ...).

**Comment by mrocklin** [comment]
@shoyer this issue may interest you

**Comment by mrocklin** [comment]
Here is a concrete suggestion to serve as a first pass:

We replace ShareDict with `HighLevelGraph(Mapping)` (or some better name).  This has the same structure of a dict of mappings, keyed by level/stage name.  

```python
>>> graph = x.__dask_graph__()
>>> graph.layers
{'load': {('load', 0): ..., ('load', 1): ...},
 'process': {('process', 0), ...},
 ...
```

These layers are also mappings.  In the example above they are raw Python dictionaries but they could also be some symbolic mapping, such as we're doing in the atop fusion PR.  These might be super-generic types like `ArrayMapping` or `dict`, or they might be highly specific, as in the atop case.

```python
>>> graph.layers
{'load': <Custom Mapping Class>,
 'process': <Custom Mapping Class>,
 ...
```

Additionally, we also require dependency information between stages (this is new)

```python
>>> graph.dependencies
{'load': set(),
 'process': {'load'}
 ...  }
```

We use this class for all collections in dask/dask, but don't strictly require it.  Dask schedulers still only expect `dict`s.

Dask collections will optionally implement a new protocol `__dask_layer_names__`, that return a set of layers from which their keys come.  I believe that for all cases today in dask core collections, this set has length one, and is equivalent to just the name of the collection.

```python
>>> x.__dask_layer_names__()
{'process'}
```

This is especially useful when constructing new collections from old ones and tracking dependencies.  Here is what this would look like when writing an operation in dask dataframe or dask array.

```python
def operation(inputs):
    name = 'operation-' + tokenize(*inputs)
    layer = {(name, i): ... }  # we used to call this `dsk`
    
    # We pass in collections, let HighLevelGraph pull out layers, build dependencies, etc..
    # It only stores the layers and the dependencies though
    graph = HighLevelGraph.from_collections(name=name, layer=layer, collections=inputs)  

    return Array(graph, name, chunks, dtype)
```
   
```python 
class HighLevelGraph(collections.Mapping):
    layers: Dict[str, collections.Mapping]
    dependencies: Dict[str, Set[str]]

    def from_collections(self, name, layer, collections):
        layers = merge(c.layers for c in collections)
        layers[name] = layer

        dependencies = merge(c.dependencies for c in collections)
        dependencies[name] = set(concat(c.__dask_layer_names__() for c in collections if hasattr(c, '__dask_layer_names__')))

        self.layers = layers
        self.dependencies = dependencies
```

To be clear we're not requiring this anywhere outside of the core collections that we produce.  Dask schedulers still expect any sort of mapping, including raw dicts.

### Optimization

But within base.py functions (compute, persist, visualize, optimize) we look for this object (or some suitable method on it) and treat it specially when we optimize.  

#### Problem

Previously we assigned optimization methods to Dask collections with the `__dask_optimize__` protocol.  This proved challenging because we would generate part of the graph with one collection, but then move over to another collection before computing.

```python
x = dask.array.from_array(dataset, chunks=...)
x = x[0, :, :]  # graph layers created in dask array, want to apply dask array optimizations
df = dask.dataframe.from_dask_array(x)
df.sum().compute()  # But graph is computed with dask dataframe
```

Additionally, there are some opportunities to do optimization at the sub-mapping layer before generating the full task graph.

#### Proposal

So we move the `__dask_optimize__` method to the sub-mappings (layers) of the HighLevelGraph.  Let me expand that a bit.

```python
>>> graph.layers
{'load': <ArrayMapping>,
 'process': <ArrayMapping>,
 'convert': <DataFrameMapping>,
 'save': <DataFrameMapping>
}

>>> graph.layers['load'].__dask_optimize__
<function: optimize_dask_arrays>
```

So in base.py when we optimize we collect all `__dask_optimize__` methods on sub-mappings, and continue to call them (with some order yet to be determined) on the full graph, and all provided collections.

```python
stop = False
while not stop:  # Continue until things stop changing
    optimizations = {layer.__dask_optimize__ if hasattr(layer, '__dask_optimize__') for layer in graph.layers}  # get all optimization functions in the graph
    for opt in optimizations:  # run each of them until something happens
        new = opt(graph, collections)
        if new is not graph:  # changed?
            break
    stop = True
```

After this happens, we can convert to a dictionary and run through our low-level task graph optimizations like `cull`, `fuse`, and so on.

**Comment by shoyer** [comment]
I’m excited about this! Separating optimizations from dask collection objects is a nice bonus. I haven’t had time to think though all the details yet but so far everything looks reasonable.

**Comment by jakirkham** [comment]
A few things that may be worth revisiting in some form in this context.

1. Commutation ( https://github.com/dask/dask/issues/2648 )
2. Symmetry ( https://github.com/dask/dask/issues/3612 )
3. Branching ( https://github.com/dask/dask/issues/2579 )

**Comment by mrocklin** [comment]
As an update, the high level graph proposal has been implemented in https://github.com/dask/dask/pull/4092 and could use review by people familiar with dask's internal graph handling (nudges @jakirkham :) )

**Comment by jakirkham** [comment]
Thanks for the nudges. Will give it a look. :)

**Comment by sjperkins** [comment]
I haven't had time to drill into this but the following [use of top](https://github.com/ska-sa/codex-africanus/blob/af974cc0ac40222e60a35e2db1b93279a932acc1/africanus/rime/dask.py#L295-L421) causes the following issues with dask 0.20.0 (but its fine in 0.19.4).

```traceback
africanus/rime/tests/test_predict.py:173: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../../venv/afr/local/lib/python2.7/site-packages/dask/base.py:155: in compute
    (result,) = compute(self, traverse=False, **kwargs)
../../../../venv/afr/local/lib/python2.7/site-packages/dask/base.py:389: in compute
    dsk = collections_to_dsk(collections, optimize_graph, **kwargs)
../../../../venv/afr/local/lib/python2.7/site-packages/dask/base.py:193: in collections_to_dsk
    for opt, (dsk, keys) in groups.items()]))
../../../../venv/afr/local/lib/python2.7/site-packages/dask/array/optimization.py:38: in optimize
    dsk = optimize_atop(dsk, keys=keys)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

full_graph = <dask.sharedict.ShareDict object at 0x7f6734467e50>
keys = [('add-71b2145c6f6fbbc0c4bf33e9159a361d', 0, 0, 0), ('add-71b2145c6f6fbbc0c4bf33e9159a361d', 0, 1, 0), ('add-71b2145c6...a361d', 1, 1, 0), ('add-71b2145c6f6fbbc0c4bf33e9159a361d', 2, 0, 0), ('add-71b2145c6f6fbbc0c4bf33e9159a361d', 2, 1, 0)]

    def optimize_atop(full_graph, keys=()):
        """ High level optimization of stacked TOP layers
    
        For operations that have multiple TOP operations one after the other, like
        ``x.T + 123`` we can fuse these into a single TOP operation.  This happens
        before any actual tasks are generated, and so can reduce overhead.
    
        This finds groups of TOP operations that can be safely fused, and then
        passes them to ``rewrite_atop`` for rewriting.
    
        Parameters
        ----------
        full_graph: ShareDict
        keys: Iterable
            The keys of all outputs of all collections.
            Used to make sure that we don't fuse a layer needed by an output
    
        Returns
        -------
        sharedict : ShareDict
    
        See Also
        --------
        rewrite_atop
        """
        keep = {k[0] if type(k) is tuple else k for k in keys}
        layers = full_graph.dicts
        dependents = core.reverse_dict(full_graph.dependencies)
        roots = {k for k in full_graph.dicts
                 if not dependents.get(k)}
        stack = list(roots)
    
        out = {}
        dependencies = {}
        seen = set()
    
        while stack:
            layer = stack.pop()
            if layer in seen or layer not in layers:
                continue
            seen.add(layer)
    
            # Outer loop walks through possible output TOP layers
            if isinstance(layers[layer], TOP):
                top_layers = {layer}
                deps = set(top_layers)
                while deps:  # we gather as many sub-layers as we can
                    dep = deps.pop()
>                   if (isinstance(layers[dep], TOP) and
                            not (dep != layer and dep in keep) and  # output layer
                            layers[dep].concatenate == layers[layer].concatenate):  # punt on mixed concatenate
E                           KeyError: 'predict_vis-f323595008bcf6372de837123567e6a1'

```

I also see the following on dask 0.20.0:

```
    def test_dask_degridder_gridder():
        from africanus.filters import convolution_filter
        from africanus.gridding.simple.dask import grid, degrid
    
        da = pytest.importorskip('dask.array')
    
        row = 100
        chan = 16
        corr = (2, 2)
        nx = ny = 1024
        cell_size = 6
    
        row_chunk = 25
        chan_chunk = 4
        corr_chunk = corr
    
        vis_shape = (row, chan) + corr
        vis_chunks = (row_chunk, chan_chunk) + corr_chunk
    
        vis = (da.random.random(vis_shape, chunks=vis_chunks) +
               1j*da.random.random(vis_shape, chunks=vis_chunks))
        uvw = da.random.random((row, 3), chunks=(row_chunk, 3))
        # 4 channels of MeerKAT L band
        wavelengths = lightspeed/da.linspace(.856e9, .856e9*2, chan,
                                             chunks=chan_chunk)
        flags = da.random.randint(0, 1, size=vis_shape, chunks=vis_chunks)
    
        weights = da.random.random(vis_shape, chunks=vis_chunks)
    
        conv_filter = convolution_filter(3, 21, "kaiser-bessel")
    
        vis_grid = grid(vis, uvw, flags, weights, wavelengths, conv_filter,
>                       cell_size, ny, nx)

africanus/gridding/simple/tests/test_simple_gridder.py:243: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
africanus/gridding/simple/dask.py:51: in grid
    dtype=vis.dtype)
../../../../venv/afr/local/lib/python2.7/site-packages/dask/array/top.py:482: in atop
    dsk = _top(func, out, out_ind, *argindsstr, numblocks=numblocks, **kwargs2)
../../../../venv/afr/local/lib/python2.7/site-packages/dask/array/top.py:96: in _top
    kwargs = subs(kwargs, sub)
../../../../venv/afr/local/lib/python2.7/site-packages/dask/array/top.py:22: in subs
    return type(task)([subs(x, substitution) for x in task])
../../../../venv/afr/local/lib/python2.7/site-packages/dask/array/top.py:22: in subs
    return type(task)([subs(x, substitution) for x in task])
../../../../venv/afr/local/lib/python2.7/site-packages/dask/array/top.py:22: in subs
    return type(task)([subs(x, substitution) for x in task])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

task = ConvolutionFilter(half_sup=3, oversample=21, full_sup_wo_padding=7, full_sup=9...8.85193423e-11, ...,
        8.85193423e-11, 6.05045620e-11, 4.05508109e-11]])), substitution = {}

    def subs(task, substitution):
        """ Create a new task with the values substituted
    
        This is like dask.core.subs, but takes a dict of many substitutions to
        perform simultaneously.  It is not as concerned with micro performance.
        """
        if isinstance(task, dict):
            return {k: subs(v, substitution) for k, v in task.items()}
        if isinstance(task, (tuple, list, set)):
>           return type(task)([subs(x, substitution) for x in task])
E           TypeError: __new__() takes exactly 7 arguments (2 given)
```

**Comment by jakirkham** [comment]
Thanks for playing with it, @sjperkins. The changes to `top` is probably the one piece of the PR that I didn't read too closely (at least on a first pass). If you have time to read it a bit more closely (and maybe make suggestions), am sure you're feed back would be appreciated.

**Comment by jakirkham** [comment]
IDK if it would be possible here, but it could be nice to handle fusing of `map_overlap` calls as well.

**Comment by mrocklin** [comment]
@sjperkins given that that seems to be an issue with the current release and not this issue I recommend raising a separate issue.  As always, if you're able to provide a minimal reproducible example that greatly accelerates things.

**Comment by sjperkins** [comment]
@mrocklin @jakirkham See https://github.com/dask/dask/issues/4176.
